### PR TITLE
Network graph new

### PR DESCRIPTION
**Pull request by pulkit10251**
This PR fixes/resolves issue #805. One can now visualize Network graphs, nodes, and edges using the viz.graph() function. It's now highly interactive and provide some customization options also
## Description

In this pull request I have made changes at the following places. 
. server.py :
Added code in def window() function to handle the opts parameters. Also added downloadable scripts for these libraries:

d3.v3.min.js (for implementing d3 based network visualization)
d3-selection-multi.v1.js (to set multiple CSS properties for d3 graph)
saveSvgAsPng.js (to be able to download SVG image in PNG format)
2. Network.css:
Added CSS for NetworkPane.js.

3. Index.html:
Added scripts and css for the above-downloaded libraries and newly created css file.

4. NetworkPane.js:
Created a code to visualize graphs in javascript. It was done using the d3 library.

5. init.py
Created a function def graph() to get data from the user and validate the user input using the library "cerberus" and send the validated data to the server.

6. init.pyi:
Added definition for the newly created graph.

7. setup.py:
Added the "cerberus" library in the requirements.

8. demo.py:
Added demo code for graph().

## Motivation and Context

Till this pull request, it was not possible to visualize graphs in visdom, though it was possible to display non-interactive image using networkx library, there is no way to display network-graph, such that it remains interactive with the user. Thus this pull request is a feature addition to visdom.

Yes, it fixes this issue. this pull request is made in response to issue #805.
## How Has This Been Tested?

I have run demo.py to test the functionality and it is working fine.
## Screenshots (if appropriate):

graph
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] For JavaScript changes, I have re-generated the minified JavaScript code.

### CODE REVIEW

**Comment by pulkit10251** [comment]
Hi @dizcza  and @JackUrb . I hope you are both are doing great. Please let me know if something needs to be updated or changed. This time I have fixed the resizing issue and added new input style (based on my discussion with @dizcza.
Thanks

File: py/visdom/__init__.py
Comment (dizcza) [inline py/visdom/__init__.py:2393]: That's a bit awkward: you require "edges" as the input, then create a graph from it to finally extract the edges back again.
If you stick to networkx, wouldn't make it convenient to require the graph itself as the input? I mean, you already require the networkx to be installed.

**Review by dizcza — APPROVED** [review]
I'm fine with the overall implementation.
Accepted.

The only question is: you're trying to add a file with 9000+ lines. Do you use `package-lock.json` somewhere? How large the file is in MB? Maintainers, is there a limit on additional files to be included in the package?

File: py/visdom/__init__.py
Comment (dizcza) [reply py/visdom/__init__.py:2393]: This is a minor comment anyway.

Comment (pulkit10251) [reply py/visdom/__init__.py:2393]: > I'm fine with the overall implementation. Accepted.
> 
> The only question is: you're trying to add a file with 9000+ lines. Do you use `package-lock.json` somewhere? How large the file is in MB? Maintainers, is there a limit on additional files to be included in the package?

Hi @dizcza. I have deleted all the extra files in my latest commit. Now everything is fine. I have tested it in a new virtual env.  and for the second part, the main thought behind this is that I didn't want the user to download any extra library besides visdom. The reason for using networkx again is only to get nodes out of edges ( which i could have done using set and for loop), and also i was checking if there is a mech inside networkx to check for errors when given wrong input this way i wanted to reduce the work for validation.

**Comment by pulkit10251** [comment]
HI @dizcza and @JackUrb, please let me know if something needs to change or something is pending.

File: py/visdom/__init__.py
Comment (dizcza) [reply py/visdom/__init__.py:2393]: That's fine. You or me, or anyone else could improve this in the future once we understand how to do it "better".

File: setup.py
Comment (dizcza) [inline setup.py:49]: well, since you add networkx into the requirements list you don't need to do
```python
        try:
            import networkx as nx
        except:
            raise RuntimeError(
                "networkx must be installed to plot Graph figures")
```
instead, import networkx in the file header.

But again, someone will fix this in the upcoming pull requests on this matter.

If I were a maintainer, I'd already start splitting the requirements into hard requirements and requirements-extra.

**Review by dizcza — APPROVED** [review]
I've already approved this PR. Waiting for @JackUrb or other devs to merge it.

**Comment by dizcza** [comment]
@pulkit10251 please link the issue https://github.com/fossasia/visdom/issues/805 in this PR description. And close the previous PR.

For example, I always start a PR with "This PR fixes/resolves the issue \<link\> ..."

Once this PR is merged, my original issue will be closed automatically.

**Comment by pulkit10251** [comment]
> @pulkit10251 please link the issue #805 in this PR description. And close the previous PR.
> 
> For example, I always start a PR with "This PR fixes/resolves the issue <link> ..."
> 
> Once this PR is merged, my original issue will be closed automatically.

Added.

**Comment by pulkit10251** [comment]
> I've already approved this PR. Waiting for @JackUrb or other devs to merge it.

Thanks, @dizcza. I am glad I have come to my first open-source contribution and your comments helped a lot!

File: example/demo.py
Comment (JackUrb) [inline example/demo.py]: nit - unnecessary whitespace change.

File: py/visdom/__init__.pyi
Comment (JackUrb) [inline py/visdom/__init__.pyi]: nit: missing newline

File: py/visdom/server.py
Comment (JackUrb) [inline py/visdom/server.py:1667]: Nit - it would be reasonable to include a comment above these scripts about what they're imported for.

File: py/visdom/server.py
Comment (JackUrb) [inline py/visdom/server.py]: Nit: unnecessary spacing.

File: py/visdom/static/css/network.css
Comment (JackUrb) [inline py/visdom/static/css/network.css]: nit: newline

**Review by JackUrb — APPROVED** [review]
Wow this is a great PR - thanks @pulkit10251 for making this happen, and @dizcza for the feedback along the way. The inclusion of docs and a demo will definitely lead to usage, especially considering this was a fairly long-requested feature for the project.

I have a few tiny nitpick comments about whitespace, but just in the interest of keeping the codebase neat.

Agreed on the overall discussion on visdom having optional requirements, but I'm not sure we've hit that point just yet. - One overall thought is that having optional requirements would lead us to also wanting to handle JS downloads separately depending on which extensions are installed, but we can cross that bridge when people begin raising it as an issue.

Happy to merge in after the small whitespace nits are covered!

File: example/demo.py
Comment (pulkit10251) [reply example/demo.py]: Done!

File: py/visdom/__init__.pyi
Comment (pulkit10251) [reply py/visdom/__init__.pyi]: Done!

File: py/visdom/server.py
Comment (pulkit10251) [reply py/visdom/server.py:1667]: Done!

File: py/visdom/server.py
Comment (pulkit10251) [reply py/visdom/server.py]: Done!

File: py/visdom/static/css/network.css
Comment (pulkit10251) [reply py/visdom/static/css/network.css]: Done!

**Comment by pulkit10251** [comment]
> Wow this is a great PR - thanks @pulkit10251 for making this happen, and @dizcza for the feedback along the way. The inclusion of docs and a demo will definitely lead to usage, especially considering this was a fairly long-requested feature for the project.
> 
> I have a few tiny nitpick comments about whitespace, but just in the interest of keeping the codebase neat.
> 
> Agreed on the overall discussion on visdom having optional requirements, but I'm not sure we've hit that point just yet. - One overall thought is that having optional requirements would lead us to also wanting to handle JS downloads separately depending on which extensions are installed, but we can cross that bridge when people begin raising it as an issue.
> 
> Happy to merge in after the small whitespace nits are covered!

Hi @JackUrb. I have merged these minor changes as you specified. Thanks

**Review by JackUrb — APPROVED** [review]

**Pull request merged**
